FlashLFQ/Bayesian Protein Quantification/ProteinQuantificationEngine.cs

```diff
@@ -0,0 +1,604 @@
+﻿using BayesianEstimation;
+using MathNet.Numerics.Random;
+using MathNet.Numerics.Statistics;
+using MzLibUtil;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace FlashLFQ
+{
+    /// <summary>
+    /// This is the Bayesian protein quantification engine used by FlashLFQ. It uses a Markov Chain Monte Carlo (MCMC)
+    /// algorithm to estimate the fold-change of a protein between conditions, given its constituent peptides fold-change measurements.
+    /// This estimation is performed by fitting Student's t-distribution to the peptide fold-change data.
+    /// 
+    /// The posterior error probability (PEP) of a protein is the probability that the protein's fold-change is below a certain value ("cutoff").
+    /// This "cutoff" can be defined by the user or can be estimated from the data using an "experimental null" strategy.
+    /// The PEP is estimated using a "skeptical prior probability", which means that the initial assumption (given no data) is that the 
+    /// protein is not changing in abundance between conditions. The number of data points (peptide fold-changes for a protein), the 
+    /// magnitude of the fold-change relative to the cutoff, and the precision of the data points all factor in to this PEP calculation.
+    /// 
+    /// In other words, the default assumption is that the protein is not changing; it must be "proven" to be changing by examining the evidence.
+    /// 
+    /// See the BayesianEstimation project for details of the MCMC algorithm.
+    /// </summary>
+    public class ProteinQuantificationEngine
+    {
+        private readonly FlashLfqResults results;
+        private readonly int MaxThreads;
+        private readonly bool UseSharedPeptides;
+        private readonly bool PairedSamples;
+        private readonly string BaseCondition;
+        private readonly double? FoldChangeCutoff;
+        private readonly int BurnInSteps;
+        private readonly int NSteps;
+        private readonly MersenneTwister Rng;
+        private bool conditionsAreDefined;
+        private Dictionary<ProteinGroup, List<Peptide>> ProteinsWithConstituentPeptides;
+        private Dictionary<(Peptide, string, int), double> PeptideToSampleQuantity;
+        private Dictionary<ProteinGroup, double> ProteinToBaseConditionIntensity;
+
+        /// <summary>
+        /// Constructs the protein quantification engine. 
+        /// 
+        /// "baseCondition" refers to the condition that the protein fold-changes could be calculated relative to.
```

Comment:
Isn't this kind of arbitrary? Particularly if you're doing multiple comparisons (i.e. there are four different groups (A, B, C, D) and you want to compare against each of them (AvsB, AvsC, AvsD, BvsC, BvsD, CvsD). Wouldn't this clump them together so that it'd be comparing AvsBCD?

Reply:
sort of. it depends on the experimental design. typically there's some condition you're calculating intensities relative to, like "normal vs. tumor". the engine is designed with this in mind. You could potentially construct an all vs. all statistical analysis, but it is not necessary to change the engine to accomplish this. you could just call this engine with different base conditions as needed.